clone-master-clean-up.1
.\"/* 
.\" * All rights reserved
.\" * Copyright (c) 2016 SUSE LINUX GmbH, Nuernberg, Germany.
.\" * Authors: Howard Guo <hguo@suse.com>
.\" *
.\" * This program is free software; you can redistribute it and/or
.\" * modify it under the terms of the GNU General Public License
.\" * as published by the Free Software Foundation; either version 2
.\" * of the License, or (at your option) any later version.
.\" *
.\" * This program is distributed in the hope that it will be useful,
.\" * but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" * GNU General Public License for more details.
.\" */
.\" 
.TH clone-master-clean-up "1" "September 2024" "" "Clean-Up For Cloning Preparation"
.SH NAME
clone\-master\-clean\-up [OPTIONS] - Clean up a system for cloning preparation.

.SH OPTIONS
.IP "\fB\-n, \-\-dont\-ask\fR"
Do not ask any questions. Run the program with default values even it is not recommended. The root password will also be retained.
.IP "\fb\-f, \-\-dont\-change\-fstab\fR"
Do not swap UUID and label into device name in fstab.
.IP "\fb\-h, \-\-help\fR"
Print a usage message to stdout and exit successfully.

.SH DESCRIPTION
Clean up a system and delete all usage history to prepare it for being cloned.

How to properly clean up a system to create a golden image is highly dependent on installed software, machine configuration, and how the image is supposed to be used. 

.RS 5
1. A golden image for fast and easy way to deploy a freshly installed base operating system (a so-called plain vanilla image).
.PP
2. A golden image that is used during the life cycle management of a system or solution landscape.
.PP
.RE
Life-cycle management images normally include custom additions beyond the base operating system which are needed on all rolled-out systems. This can include additional software, e.g. for backup, hardware support, or monitoring, and system configuration, e.g. the creation of additional local users and changes to configuration files such as 'sudoers' or 'pam'.
.br
Additionally, with the use case of a life cycle management image, there will be a very intensive test cycle of newly installed software and configuration before the golden image is created and rolled out. Such testing can lead a lot of additional log files and temporary data.
.PP
This means that the parts of the operating system which need to be cleaned up are different depending on the use case and that it can be counterproductive to remove some special files or directories.
.PP
For the use case of creating an clean operating system base installation, the standard scope of 'clone-master-clean-up' is generally sufficient.
To allow customization for this use case, use the configuration file \fB/etc/sysconfig/clone-master-clean-up\fR.
.br
Here, you can configure the following behavior:
.RS 5
/etc/sudoers is restored to original, and only root is allowed to run sudo.
.RS 5
default: no
.RE
clean up user ids >= 1000
.RS 5 
default: no
.RE
clean up all zypper repositories
.RS 5
default: no
.RE
remove btrfs snapshots of root filesystem
.RS 5
default: yes
.RE
(If the system is running on EC2) Allow ec2-user to run sudo password-less.
.RS 5
default: no
.RE
.RE
.PP
For the use case of a life cycle management image, it's impossible to write a standard script which will fit all requirements. To enable this use case, clone-master-clean-up allows including a drop-in file containing paths to files and directories that are deleted in addition to doing the standard clean-ups.
.PP
For more information and examples, see \fB/usr/share/clone-master-clean-up/custom_remove.template\fR. Please copy the template file \fB/usr/share/clone-master-clean-up/custom_remove.template\fR to  \fB/etc/clone-master-clean-up/custom_remove\fR and un-comment the parts you need or add additional parts which will fit your needs.
.PP
If the file \fB'/etc/clone-master-clean-up/custom_remove'\fR exists, all included files and directories will be picked up line by line and additionally removed from the golden image.

.SH "CLEANED UP ITEMS"
The program asks for confirmation before proceeding with cleanup. If you proceed by typing "YES" and enter, the following locations/items will be cleaned up (or deleted):

.IP \[bu]
SUSE registration, all software repositories.
.IP \[bu]
SSH host keys, root user SSH keys, root user authorized keys, root user shell history.
.IP \[bu]
User mails and user cron jobs.
.IP \[bu]
Temporary files under /tmp and /var/tmp.
.IP \[bu]
Log files and archived log files under /var/log.
.IP \[bu]
Auto-generated HANA firewall configuration script.
.IP \[bu]
System random seeds.
.IP \[bu]
Systemd journal.
.IP \[bu]
machine ID.
.IP \[bu]
Salt client ID.
.IP \[bu]
osad authentication configuration file and the system ID.
.IP \[bu]
Domain and host names.
.IP \[bu]
/etc/hostname, /etc/hosts, /etc/resolv.conf are restored to original.
.IP \[bu]
Persistent network interfaces (auto-generated 70-persistent-net.rules).
.IP \[bu]
YaST Firstboot is enabled if there is a firstboot.xml.
.IP \[bu]
Give root user a new password, optionally.
.IP \[bu]
Zypper anonymous ID
.IP \[bu]
Clean up postfix
.IP \[bu]
Swap the uuid strings with dev strings in /etc/fstab
.IP \[bu]
Clean up network files (except interfaces using dhcp boot protocol)
.IP \[bu]
Clean up collectd
.IP \[bu]
Clean up /root
.IP \[bu]
Clean up cache, crash and coredump

.SH FILES
.PP
\fI/etc/sysconfig/clone-master-clean-up\fR
.RS 4
Defines the behavior of clone-master-clean-up.
.RE
.PP
\fI/usr/share/clone-master-clean-up/custom_remove.template\fR
.RS 4
Template for a vendor/customer-specific drop-in file. Defines additional files and directories to clean up.
.RE
.PP
\fI/etc/clone-master-clean-up/custom_remove\fR
.RS 4
\'Operation-ready\' copy of the template file. If this file exists, clone-master-clean-up will remove files and directories that are listed in it.
.RE

.SH AUTHOR
.NF
Howard Guo <hguo@suse.com>, Angela Briel <abriel@suse.com>, Peter Varkoly <varkoly@suse.com>


clone-master-clean-up.sh
#!/bin/bash

# The script cleans up a system for cloning preparation.
# Author: Peter Varkoly <varkoly@suse.com>
# Author: Howard Guo <hguo@suse.com>

set -e
shopt -s nullglob
# bsc#1092378
DROP_IN_FILE=/etc/clone-master-clean-up/custom_remove
SYSCONF_FILE=/etc/sysconfig/clone-master-clean-up

err_exit() {
  echo "Sorry! An error occured on line $1, the clean up routine did not complete successfully."
  exit 1
}
trap 'err_exit $LINENO' ERR

while [[ $# -gt 0 ]]; do
    case $1 in
        -n|--dont-ask)
            DONT_ASK="YES"
            shift
            ;;
        -f|--dont-change-fstab)
            DONT_CHANGE_FSTAB="YES"
            shift
            ;;
        -h|--help)
            echo -e 'Usage: clone-master-clean-up.sh [OPTION..]  \n' \
                    '  -n, --dont-ask           Do not ask any questions. Run the program with default values even it is not recommended.\n' \
                    '                           The root password will also be retained.\n' \
                    '  -f, --dont-change-fstab  Do not swap UUID and label into device name in fstab.\n'
            exit
    esac
done

[ "$UID" != "0" ] && echo 'Please run this program as root user.' && exit 1

if [ "${DONT_ASK}" != "YES" ]; then
    echo -e 'The script will delete root SSH keys, log data, and more.\n' \
         'WARNING: This should only be used on a pristine system\n' \
         'WARNING: with no populated /home directories!\n' \
         'Type YES and enter to proceed.'
    read -r answer
    [ "$answer" != "YES" ] && exit 1

    if [ -n "$(echo /home/*/.ssh/* /home/*/.*_history)" ]; then
        echo -e 'There seem to be populated /home directories on this system\n' \
                'Cloning such systems is not recommended.\n' \
                'Type YES if you still would like to proceed.'
        read answer
        [ "$answer" != "YES" ] && exit 1
    fi
fi

# source config file
if [ -r "$SYSCONF_FILE" ]; then
    . "$SYSCONF_FILE"
else
    echo 'Failed to read /etc/sysconfig/clone-master-clean-up'
    exit 1
fi

echo 'Wiping active swap devices/files (this may take a while)'
while read -r swap_name discard; do
  uuid=$(env $(blkid -o export "$swap_name") printenv UUID)
  echo "Turning off swap device/file $swap_name (UUID $uuid)"
  swapoff "$swap_name"
  echo "Zero-overwriting $swap_name..."
  shred -n1 --random-source=/dev/zero "$swap_name"
  mkswap "$swap_name" -U "$uuid"
done < <(tail -n+2 /proc/swaps)

echo 'Removing system registration information and zypper repositories'
> /etc/SUSEConnect
find /etc/zypp \( -iname 'suse*' -o -iname 'scc*' \) -delete

echo "Removing zypper anonymous ID"
rm -rf /var/lib/zypp/AnonymousUniqueId

echo 'Removing SSH host keys, root user SSH keys, authorized keys, and shell history'
rm -rf /etc/ssh/ssh_host*key* /root/.ssh/*  &> /dev/null

echo 'Removing all mails and cron-jobs'
rm -rf /var/spool/mail/*
rm -rf /var/spool/cron/{lastrun,tabs}/*

echo "Clean up postfix"
for i in /var/spool/postfix/{active,corrupt,deferred,hold,maildrop,saved,bounce,defer,flush,incoming,trace}; do
    if [ -d "$i" ]; then
        # descend following symlink and check if it was symlink, if not, recursively delete entries in this directory. 'rm -rf' doesn't follow symlinks.
        cd -P "$i"
        [ "$i" != "$PWD" ] && continue
        info=( $(stat --printf="%u %g" ".") )
        owner=${info[0]}
        group=${info[1]}
        setpriv --clear-groups --reuid "$owner" --regid "$group" rm -rf ./*
    fi
done

echo 'Removing all temporary files'
rm -rf /tmp/* /tmp/.* /var/tmp/* /var/tmp/.* &> /dev/null || true

echo 'Removing log archives'
find /var/log \( -iname '*.old' -o -iname '*.xz' -o -iname '*.gz' \) -delete
echo 'Clearing log files'
find /var/log -type f -exec truncate -s 0 {} \;

echo 'Clearing HANA firewall script'
rm -rf /etc/hana-firewall.d/generated_hana_firewall_script

echo 'Removing random seeds'
for seed in /var/lib/systemd/random-seed /var/lib/misc/random-seed; do
[ -f "$seed" ] && rm -f "$seed"
done

echo 'Clearing systemd journal'
pushd /etc/systemd > /dev/null
[ -e journald.conf ] && cp journald.conf journald.conf.bak
echo -e '\nSystemMaxUse=1K' >> journald.conf
systemctl restart systemd-journald
rm journald.conf
[ -e journald.conf.bak ] && mv journald.conf.bak journald.conf
popd > /dev/null

echo 'Clearing machine ID file'
# on distributions that support systemd
truncate -s 0 /etc/machine-id
# on distributions that do not support systemd
[ ! -f /etc/machine-id ] && truncate -s 0 /var/lib/dbus/machine-id

echo 'Removing Salt client ID'
[ -f /etc/salt/minion_id ] && rm -f /etc/salt/minion_id

echo 'Removing osad authentication configuration file and the system ID'
rm -f /etc/sysconfig/rhn/{osad-auth.conf,systemid}

echo 'Removing domain name and set host name from DHCP in network config'
sed -i 's/^NETCONFIG_DNS_STATIC_SEARCHLIST=.*$/NETCONFIG_DNS_STATIC_SEARCHLIST=""/g' /etc/sysconfig/network/config
sed -i 's/^DHCLIENT_SET_HOSTNAME=.*$/DHCLIENT_SET_HOSTNAME="yes"/g' /etc/sysconfig/network/dhcp

echo 'Removing persistent network interfaces'
netudev=/etc/udev/rules.d/70-persistent-net.rules
if [ -f "$netudev" ]; then
  if grep -i 'automatically generated' "$netudev" &> /dev/null; then
    rm "$netudev"
  fi
fi

echo 'Restoring initial system-wide network config'
truncate -s 0 /etc/{hostname,hosts,resolv.conf}
cat << EOF > /etc/hosts
127.0.0.1       localhost
::1             localhost ipv6-localhost ipv6-loopback
fe00::0         ipv6-localnet
ff00::0         ipv6-mcastprefix
ff02::1         ipv6-allnodes
ff02::2         ipv6-allrouters
ff02::3         ipv6-allhosts
EOF

echo 'Enabling YaST Firstboot if necessary'
[ -e /etc/YaST2/firstboot.xml ] && touch /var/lib/YaST2/reconfig_system


if [ "$CMCU_RSNAP" = "yes" ]; then
  if [ -d /.snapshots ]; then
    echo "Removing all pre/post btrfs snapshots from /.snapshot"
    presnapshots=$(dbus-send --type=method_call --system --print-reply \
                             --dest=org.opensuse.Snapper \
                             /org/opensuse/Snapper \
                             org.opensuse.Snapper.ListSnapshots string:root \
                             2>/dev/null | awk -- "
BEGIN {arr=0; cnt=0; u2=0; u4=0; del=0}
/array \[/ {arr++}
/struct {/ {if (arr==1) cnt++}
/}/ {if(arr==1&&--cnt==0){if(del==1) print id ;del=0;u4=0;u2=0}}
/\]/ {arr--}
# Don't delete current snapshot
/string "current"/ {if (arr==1 && cnt==1) del=0}
# ID: 1st uint32 value of each top struct in top array
/uint32/ {if (arr==1 && cnt==1) if (++u4==1)id=\$2; else if (u4==2)lst=\$2}
# Type: 1st uint16 value of each top struct in top array
/uint16/ {if (arr==1 && cnt==1){if (++u2==1) {if (\$2==1 ){del=1}}}}
")
    for i in $presnapshots
    do
       /usr/bin/snapper delete --sync $i
    done
    postsnapshots=$(dbus-send --type=method_call --system --print-reply \
                             --dest=org.opensuse.Snapper \
                             /org/opensuse/Snapper \
                             org.opensuse.Snapper.ListSnapshots string:root \
                             2>/dev/null | awk -- "
BEGIN {arr=0; cnt=0; u2=0; u4=0; del=0}
/array \[/ {arr++}
/struct {/ {if (arr==1) cnt++}
/}/ {if(arr==1&&--cnt==0){if(del==1) print id ;del=0;u4=0;u2=0}}
/\]/ {arr--}
# Don't delete current snapshot
/string "current"/ {if (arr==1 && cnt==1) del=0}
# ID: 1st uint32 value of each top struct in top array
/uint32/ {if (arr==1 && cnt==1) if (++u4==1)id=\$2; else if (u4==2)lst=\$2}
# Type: 1st uint16 value of each top struct in top array
/uint16/ {if (arr==1 && cnt==1){if (++u2==1) {if (\$2==2 ){del=1}}}}
")
    for i in $postsnapshots
    do
       /usr/bin/snapper delete --sync $i
    done
  fi
fi

if [ "$CMCU_ZYPP_REPOS" = "yes" ]; then
    echo "Clean up all zypper repositories"
    rm -rf /etc/zypp/repos.d/*
fi
if [ "$CMCU_SUDOERS" = "yes" ]; then
    echo 'Restoring initial sudoers settings'
    cat << EOF > /etc/sudoers
Defaults always_set_home
Defaults env_reset
Defaults env_keep = "LANG LC_ADDRESS LC_CTYPE LC_COLLATE LC_IDENTIFICATION LC_MEASUREMENT LC_MESSAGES LC_MONETARY LC_NAME LC_NUMERIC LC_PAPER LC_TELEPHONE LC_TIME LC_ALL LANGUAGE LINGUAS XDG_SESSION_COOKIE"
Defaults secure_path="/usr/sbin:/usr/bin:/sbin:/bin"
Defaults !insults
Defaults targetpw

ALL ALL=(ALL) ALL
root ALL=(ALL) ALL
EOF
fi

if [ "${DONT_ASK}" != "YES" ]; then
    echo 'Would you like to give root user a new password? Type YES to set a new password, otherwise simply press Enter.'
    read -r answer
    [ "$answer" == "YES" ] && passwd root
fi
if [ "$CMCU_EC2" = "yes" ]; then
    if curl --connect-timeout 3 '169.254.169.254/latest' &> /dev/null; then
       echo 'EC2 specific: allowing ec2-user to run sudo'
       echo 'ec2-user ALL=(ALL) ALL' >> /etc/sudoers
       echo 'Note to EC2 user:'
       echo 'This instance will no longer be accessible after a reboot due to removal of all authorized SSH keys.'
       echo "New SSH key will be installed automatically if you create a new instance from this instance's image."
    fi
fi

if [ "$CMCU_USERIDS" = "yes" ]; then
    echo "clean up user ids >= 1000"
    for i in $(awk -F ":" '$3 >= 1000 && $1 !~ /nobody/ {print $1}' /etc/passwd); do
        userdel -r "$i"
    done
fi

if [ "${DONT_CHANGE_FSTAB}" != "YES" ]; then
    echo "swap the uuid strings with dev strings in /etc/fstab"
    > /tmp/fstab.tmp
    while read -r disk remain; do
        case "$disk" in
        UUID=*)
            uuid=${disk#UUID=}
            new_disk=$(/usr/sbin/blkid -U "$uuid")
        ;;
        LABEL=*)
            label=${disk#LABEL=}
            new_disk=$(/usr/sbin/blkid -L "$label")
        ;;
        *)
            new_disk="$disk"
        ;;
        esac
        echo "$new_disk $remain" >> /tmp/fstab.tmp
    done < /etc/fstab
    if [ -s /tmp/fstab.tmp ]; then
        cp /tmp/fstab.tmp /etc/fstab
    fi
    rm -rf /tmp/fstab.tmp
fi

echo "Clean up network files (except interfaces using dhcp boot protocol)"
# additional files like bondig interfaces or vlans can be found in
# /usr/share/clone-master-clean-up/custom_remove.template
for intf in /etc/sysconfig/network/ifcfg-eth*; do
    bprot=$(grep "^BOOTPROTO=" "$intf" | sed "s/^BOOTPROTO=//")
    if ! [[ "$bprot" =~ dhcp ]]; then
        rm -rf "$intf"
    fi
done
if [ -d /var/lib/wicked ]; then
    echo "Clean up persistent network data"
    rm -rf /var/lib/wicked/*
fi

echo "Clean up collectd"
rm -rf /var/lib/collectd/WebYaST/*

echo "Clean up /root"
rm -rf /root/.viminfo /root/.bash_history /root/.Xauthority /root/.ICEauthority /root/.xsession-errors* /root/.dbus/session-bus/* /root/.config/* /root/.cache/*

echo "Clean up cache, crash and coredump"
rm -rf /var/cache/*/* /var/crash/* /var/lib/systemd/coredump/*

# checking for 'drop-in' file
if [ -r "$DROP_IN_FILE" ]; then
    echo "Remove files and directories found in the drop-in file '$DROP_IN_FILE'"
    while read line; do
        [[ "$line" =~ ^# ]] && continue
        if [ -f "$line" ]; then
            rm "$line"
        elif [ -d "$line" ]; then
            rm -rf "$line"
        else
            echo "Error: '$line' has wrong format. At the moment only files or directories are supported."
        fi
    done < $DROP_IN_FILE
fi

if [ -e /etc/iscsi/initiatorname.iscsi ]; then
    echo 'Clean up initiatorname.iscsi'
    sed -i '/^[^#]/d' /etc/iscsi/initiatorname.iscsi
fi

echo 'Finished. The system is now sparkling clean. Feel free to shut it down and image it.'

